swift/apple/FirezoneKit/Sources/FirezoneKit/Views/MenuBar.swift

```diff
@@ -371,38 +360,40 @@
       }()
     }
 
-    private func handleMenuVisibilityOrStatusChanged() {
-      let status = appStore.tunnelStore.status
-      if isMenuVisible && status == .connected {
-        appStore.tunnelStore.beginUpdatingResources()
+    private func resourceMenuTitle(_ newResources: [Resource]?) -> String {
+      guard let newResources = newResources else { return "Loading Resources..." }
+
+      if newResources.isEmpty {
+        return "No Resources"
       } else {
-        appStore.tunnelStore.endUpdatingResources()
+        return "Resources"
       }
     }
 
-    private func setOrderedResources(_ newOrderedResources: [DisplayableResources.Resource]) {
-      if resourcesTitleMenuItem.isHidden && resourcesSeparatorMenuItem.isHidden {
-        guard newOrderedResources.isEmpty else {
-          return
-        }
+    private func handleMenuVisibilityOrStatusChanged() {
+      if isMenuVisible && tunnelStore.status == .connected {
+        tunnelStore.beginUpdatingResources()
+      } else {
+        tunnelStore.endUpdatingResources()
       }
-      let diff = newOrderedResources.difference(
-        from: self.orderedResources,
-        by: { $0.name == $1.name && $0.location == $1.location }
+    }
+
+    private func populateResourceMenu(_ newResources: [Resource]?) {
+      // the menu contains other things besides resources, so update it in-place
+      let diff = (newResources ?? []).difference(
+        from: resources ?? [],
+        by: { $0.name == $1.name && $0.address == $1.address }
       )
-      let baseIndex = menu.index(of: resourcesTitleMenuItem) + 1
+      let index = menu.index(of: resourcesTitleMenuItem) + 1
       for change in diff {
         switch change {
         case .insert(let offset, let element, associatedWith: _):
-          let menuItem = createResourceMenuItem(title: element.name, submenuTitle: element.location)
-          menu.insertItem(menuItem, at: baseIndex + offset)
-          orderedResources.insert(element, at: offset)
+          let menuItem = createResourceMenuItem(title: element.name, submenuTitle: element.address)
+          menu.insertItem(menuItem, at: index + offset)
         case .remove(let offset, element: _, associatedWith: _):
-          menu.removeItem(at: baseIndex + offset)
-          orderedResources.remove(at: offset)
+          menu.removeItem(at: index + offset)
```

Comment:
Hm, the Apple client does a diff and mutates the menu instead of blowing it up and regenerating it all? Hope it doesn't de-sync.

Reply:
Yeah it's been pretty stable so far. I think this can be redone though as part of #3514